bftengine/src/bftengine/SeqNumInfo.cpp

```diff
@@ -266,7 +289,135 @@ Time SeqNumInfo::getTimeOfFirstRelevantInfoFromPrimary() const { return firstSee
+}
+
+FullCommitProofMsg* SeqNumInfo::getFastPathFullCommitProofMsg() const {
+  auto optimistic = fastPathOptimisticCollector->getMsgWithValidCombinedSignature();
+  auto threshold = fastPathThresholdCollector->getMsgWithValidCombinedSignature();
+  ConcordAssert(!(optimistic && threshold));
+  return optimistic ? optimistic : threshold;
+}
+
+void SeqNumInfo::setFastPathTimeOfSelfPartialProof(const Time& t) { fastPathTimeOfSelfPartialProof = t; }
+
+Time SeqNumInfo::getFastPathTimeOfSelfPartialProof() const { return fastPathTimeOfSelfPartialProof; }
+
+bool SeqNumInfo::addFastPathSelfPartialCommitMsgAndDigest(PartialCommitProofMsg* m, Digest& commitDigest) {
+  ConcordAssert(m != nullptr);
+  const ReplicaId myId = m->senderId();
+  ConcordAssert(myId == replica->getReplicasInfo().myId());
+  ConcordAssert(hasMatchingPrePrepare(m->seqNumber()));
+  ConcordAssert(!hasFastPathFullCommitProof());
+  ConcordAssert(getSelfCommitPartialMsg() == nullptr);
+
+  bool result = false;
+  switch (m->commitPath()) {
+    case CommitPath::OPTIMISTIC_FAST:
+      fastPathOptimisticCollector->setExpected(m->seqNumber(), m->viewNumber(), commitDigest);
+      result = fastPathOptimisticCollector->addMsgWithPartialSignature(m, myId);
+      break;
+    case CommitPath::FAST_WITH_THRESHOLD:
+      fastPathThresholdCollector->setExpected(m->seqNumber(), m->viewNumber(), commitDigest);
+      result = fastPathThresholdCollector->addMsgWithPartialSignature(m, myId);
+      break;
+    default:
+      ConcordAssert(false && "Unhandled CommitPath!");
+  }
+
+  return result;
+}
+
+bool SeqNumInfo::addFastPathPartialCommitMsg(PartialCommitProofMsg* m) {
+  ConcordAssert(m != nullptr);
+
+  const ReplicaId repId = m->senderId();
+  ConcordAssert(repId != replica->getReplicasInfo().myId());
+  ConcordAssert(replica->getReplicasInfo().isIdOfReplica(repId));
+  // PartialCommitProofMsg is allowed before a prePrepare is received
+  ConcordAssert(!prePrepareMsg || hasMatchingPrePrepare(m->seqNumber()));
+
+  if (hasFastPathFullCommitProof()) return false;
+
+  const auto* selfPartialCommitProof = getFastPathSelfPartialCommitProofMsg();
+  const CommitPath cPath = m->commitPath();
+
+  // Check different fast commit path
+  if (selfPartialCommitProof && selfPartialCommitProof->commitPath() != cPath) {
+    LOG_WARN(CNSUS,
+             "Ignoring PartialCommitProofMsg (" << CommitPathToStr(cPath) << "). Current path is "
+                                                << CommitPathToStr(selfPartialCommitProof->commitPath()));
+    return false;
+  }
+
+  auto result = false;
+  switch (cPath) {
+    case CommitPath::OPTIMISTIC_FAST:
+      result = fastPathOptimisticCollector->addMsgWithPartialSignature(m, repId);
+      break;
+    case CommitPath::FAST_WITH_THRESHOLD:
+      result = fastPathThresholdCollector->addMsgWithPartialSignature(m, repId);
+      break;
+    default:
+      ConcordAssert(false && "Unhandled CommitPath!");
```

Comment:
It seems it does, however what error handling do you suggest in this case? There are other examples where something already validated in the Validate method is a ConcordAssert - for example:
ConcordAssert(repsInfo->isIdOfPeerReplica(msgSender));
ConcordAssert(repsInfo->isCollectorForPartialProofs(msgView, msgSeqNum));
in void ReplicaImp::onMessage<PartialCommitProofMsg>(PartialCommitProofMsg *msg)

Do we:
- Rely entirely on the message validate method and omit additional error checking? 
- Keep additional error checking - either an assert or error log and if it occurs we improve the validate method?

Reply:
Basically the convention is to assert only for things that cannot escape the validate method of the message. Is it possible for a malicious replica to put a value there that could pass the validate method and trigger the assert?

Reply:
After some consideration the validate method of PartialCommitProofMsg can in fact allow invalid values for the commit path in the message that would trigger this assert. This will be addressed as part of the PR.